library/core/src/ptr/mod.rs

```diff
@@ -1901,6 +1915,8 @@ pub(crate) const unsafe fn align_offset<T: Sized>(p: *const T, a: usize) -> usiz
     ///
     /// Implementation of this function shall not panic. Ever.
     #[inline]
+    #[safety::requires(m.is_power_of_two())]
+    #[safety::requires(x < m)]
```

Comment:
I would have liked to write an `ensures` contract to check that the answer is indeed correct, i.e., that the remainder of `*result * x` divided by `m` is indeed 1. However, `x` has a multiplicative inverse mod `m` iff `gcd(x, m) == 1`, so we would have to add a precondition that `gcd(x, m) == 1`, and computing the gcd inside a contract is too time-consuming.

Reply:
Because AFAIK that only holds if the gcd of x and m is 1. The body of `mod_inv` notes that the `INV_TABLE_MOD_16` only has entries where an inverse actually exists (then does its calculations based on that). I take this comment to mean that the caller can only rely on the result to indeed be an inverse if they provide inputs for which an inverse can exist.
So ideally, I'd add a precondition like `requires(gcd(x, m) == 1` or a postcondition like `ensures(|result| if { gcd(x, m) == 1} wrapping_mul(*result, x) % m == 1 else { true })` but computing the gcd is too expensive.

Reply:
I tried adding that contract and it fails verification. Here's what concrete playback outputs:

`
#[test]
fn kani_concrete_playback_check_mod_inv_16715149060491010500() {
    let concrete_vals: Vec<Vec<u8>> = vec![
        // 0ul
        vec![0, 0, 0, 0, 0, 0, 0, 0],
        // 9223372036854775808ul
        vec![0, 0, 0, 0, 0, 0, 0, 128],
    ];
    VERIFICATION:- FAILED
Verification Time: 57.110477s

Concrete playback unit test for `ptr::verify::check_mod_inv`:
```
/// Test generated for harness `ptr::verify::check_mod_inv` that checks contract for `mod_inv_copy`
///
/// Check for `assertion`: "|result| wrapping_mul(*result, x) % m == 1"

#[test]
fn kani_concrete_playback_check_mod_inv_16715149060491010500() {
    let concrete_vals: Vec<Vec<u8>> = vec![
        // 0ul
        vec![0, 0, 0, 0, 0, 0, 0, 0],
        // 9223372036854775808ul
        vec![0, 0, 0, 0, 0, 0, 0, 128],
    ];
    kani::concrete_playback_run(concrete_vals, check_mod_inv);
}`

Reply:
How about the following:
````rust
#[requires(let d : usize = kani::any_where(|d: &usize| x % d == 0 && m % d == 0); d == 1)]
````

Reply:
Clever idea, but for some reason it's not working. I tried:

```
#[safety::requires({
  let common_divisor: usize = kani::any_where(|d: &usize| *d > 0 && x % *d == 0 && m % *d == 0);
  common_divisor == 1
})]
```

but verification still failed, and concrete playback gave me numbers that didn't seem to respect this constraint. I tried fixing `m = 8` and got this for concrete playback:

```
/// Check for `assertion`: "|result| wrapping_mul(*result, x) % m == 1"

#[test]
fn kani_concrete_playback_check_mod_inv_4047475435387855539() {
    let concrete_vals: Vec<Vec<u8>> = vec![
        // 2ul
        vec![2, 0, 0, 0, 0, 0, 0, 0],
        // 1ul
        vec![1, 0, 0, 0, 0, 0, 0, 0],
    ];
    kani::concrete_playback_run(concrete_vals, check_mod_inv);
}
```

AFAIK there's no clear way to know which one of these corresponds to `x` (I assume the other value is `result`?). So if we execute `mod_inv` for both cases:
- If `x = 1`, then `result = 1` and the postcondition holds. 
- If `x = 2`, `result = 3`, which would indeed violate the postcondition. (But this gcd precondition you suggested should prevent `x` from being 2 in the first place, since `gcd(2, 8) != 1`, i.e. `common_divisor != 1`.

All I can think is that this precondition only enforces a restriction on `common_divisor` and not `x` and `m` as we want.

Reply:
Yes -- I assume since you're asking Kani can't handle that?

Reply:
Hmm, yes, I believe you'd need quantifiers here. You'd have to say $\forall d. (d > 0 \wedge x \mod d = 0 \wedge m \mod d = 0) \rightarrow d = 1$.

Reply:
Hm, in that case I would vote that we merge it with just the preconditions for now. It's still a valuable proof, since the comment says that the implementation should never panic, so we can at least verify that even if we can't check correctness. I'll add a TODO comment to add this postcondition once quantifiers are available.

Reply:
Sounds good!